### PR TITLE
feat: Init project structure.

### PR DESCRIPTION
**Pull request by IRONICBo**
Init basic goplus community project structure. Still in progress :)

### CODE REVIEW

File: cmd/gopcomm/community_yap.gox [hunk 1]
```diff
@@ -1,10 +1,27 @@
+import "github.com/goplus/community/cmd/gopcomm/internal/config"
```
Comment (xushiwei) [inline cmd/gopcomm/community_yap.gox:1]: internal pkg can be here: "github.com/goplus/community/internal/config"

File: cmd/gopcomm/internal/config/config.gop [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+package config
+
+import (
+	"log"
+
+	"github.com/spf13/viper"
+)
+
+//Config global config instance.
+var Config *config
```
Comment (xushiwei) [inline cmd/gopcomm/internal/config/config.gop:26]: don't use global variable.

File: cmd/gopcomm/community_yap.gox (hunk 1 above)
Comment (IRONICBo) [reply cmd/gopcomm/community_yap.gox:1]: Thanks, I will move these packages.

**Comment by IRONICBo** [comment]
```bash
gopcomm % gop run .
Community server running on :8080
Visiting article 10
```

image

**Comment by IRONICBo** [comment]
Project structure.

```bash
.
├── LICENSE
├── README.md
├── cmd
│   └── gopcomm
│       ├── community_yap.gox
│       ├── config.yaml
│       ├── gop_autogen.go
│       └── yap
│           ├── article_yap.html
│           ├── edit_yap.html
│           ├── home_yap.html
│           ├── static
│           └── user_yap.html
├── docs
│   ├── account-design.md
│   ├── distribute-design.md
│   ├── markdown-design.md
│   ├── mysql-framework.md
│   ├── share-design.md
│   └── web-framework.md
├── go.mod
├── go.sum
└── internal
    ├── config
    │   └── config.gop
    └── core
        ├── article.gop
        ├── community.gop
        ├── gop_autogen.go
        └── models.gop
```

**Comment by IRONICBo** [comment]
# Design

### DOM-like Core Handlers Design

##### Top core handler

`communityHandler` define a top interface for community handler, it will return all the handlers for community.

```go
type communityHandler interface {
	GetArticleHandler() articleHandler
}
```

> Use `var _ communityHandler = (*communityHandlerImpl)(nil)` to check if the `communityHandlerImpl` implement the `communityHandler` interface.


##### Article handler

`articleHandler` define a series of method to get article in `inteface`. The `articleHandlerImpl` implement the `articleHandler` interface.

All sub handlers(like `articleHandlerImpl`, `articleHandler`) are private, only the top handler(function `New()`) is public. Developers can only use the top handler to get all the sub handlers, and use the sub handlers to get the specific operation.

```go
type articleHandler interface {
	GetArticle(id string) (Article, error)
}
```

### Config Design

`Config` is a struct to store the config of community, in `core` package, it will be used to init the community handler and all sub handlers.

In `config` package, it will be used to load the config from file or env, and used for community system.

### View Design

`View`(`router`) is used to store the view of community in `cmd/gopcomm` package, it will be used to init the web server handlers and routers.

Core web-framework is powered by `yap`(goplus web classfile), we need to define the routers to handle the request, and get top handler from `core` package.

In each router, we will get the top handler from `core` package, and use the top handler to get the sub handler, and use the sub handler to handle the request.

```go
config := &core.Config{
	core.ArticleConfig{},
}
community := core.New(config)

get "/article/:id", ctx => {
	param := ctx.param("id")
	println "Visiting article " + param

	h := community.GetArticleHandler()
	info, _ := h.GetArticle(param)

	ctx.yap "article", {
		"id": info.Id,
		"title": info.Title,
		"content": info.Content,
	}
}
```

**Pull request merged**